staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -572,8 +576,29 @@ func (w *watchCache) Replace(objs []interface{}, resourceVersion string) error {
 	w.Lock()
 	defer w.Unlock()
 
-	w.startIndex = 0
-	w.endIndex = 0
+	// We have to ensure that startIndex is never decreased.
+	// Previously we were resetting startIndex to 0 and that was exposing
+	// us to the following race:
+	// - a new Watch was registered, and watchCacheInterval is returned for it
+	//   (but it didn't yet started processing the interval)
+	// - Cacher is being reinitialized (which terminates this watcher, but
+	//   that only prevents delivering new events to it, but the ones that
+	//   are already queued may still be delievered to the watch)
+	// - watchCache is receiving new events which start to being written
+	//   starting at index 0
+	// - now the Watch from the first point starts actual processing - it
+	//   tries to load the data, but if it was conflicting with the already
+	//   overwritten indexes, it will return bad data (from the future
+	//   which then may be followed by past data for indexes which weren't
+	//   yet overwritten).
+	//
+	// Ensuring that startIndex is not decreasing, prevents us from this
+	// scenario thanks to the indexValidator which checks if the index
+	// the interval wants to download is still a valid one.
+	previousEndIndex := w.endIndex
+	w.startIndex, w.endIndex = previousEndIndex, previousEndIndex
+	w.replaceStartIndex = w.startIndex
```

Comment:
if the only use for this is to know if we've dropped any events since the last relist, a boolean would be clearer

could we:

* set `w.removedEventSinceRelist = false` here

* set it to true here:
```
		// Cache is full - remove the oldest element.
		w.startIndex++
		w.removedEventSinceRelist = true
```

* check in the case statement here:
```
	case w.listResourceVersion > 0 && !w.removedEventSinceRelist:
		// If no event was removed from the buffer since last relist, the oldest watch
		// event we can deliver is one greater than the resource version of the list.
		oldest = w.listResourceVersion + 1
```